File: fsharp/phone-number/PhoneNumber.fs
module PhoneNumber

open System

let clean (input: string): Result<uint64, string> =

    let parse (str: string) =
        let isPunctuation (c: char) = Char.IsPunctuation(c) && not("().-+".Contains(c))

        match str with
        | s when Seq.exists Char.IsLetter s -> Error "letters not permitted"
        | s when Seq.exists isPunctuation s -> Error "punctuations not permitted"
        | s -> Ok (s |> String.filter Char.IsDigit)

    let validLength (digits: string) =
        match digits with
        | s when s.Length < 10 -> Error "incorrect number of digits"
        | s when s.Length > 11 -> Error "more than 11 digits"
        | s -> Ok s

    let validCountryCode (digits: string) =
        let (|WithCountryCode|WithoutCountryCode|) (s: string) =
            match s with
            | s when s.Length = 11 -> WithCountryCode (s.[0], s.[1..])
            | s -> WithoutCountryCode s

        match digits with
        | WithCountryCode (countryCode, _) when countryCode <> '1' -> Error "11 digits must start with 1"
        | WithCountryCode (_, number) -> Ok number
        | WithoutCountryCode number -> Ok number

    let validAreaCode (digits: string) =
        match digits.[0] with
        | '0' -> Error "area code cannot start with zero"
        | '1' -> Error "area code cannot start with one"
        | _ -> Ok digits

    let validExchangeCode (digits: string) =
        match digits.[3] with
        | '0' -> Error "exchange code cannot start with zero"
        | '1' -> Error "exchange code cannot start with one"
        | _ -> Ok digits

    parse input
    |> Result.bind validLength
    |> Result.bind validCountryCode
    |> Result.bind validAreaCode
    |> Result.bind validExchangeCode
    |> Result.map uint64

File: kotlin/phone-number/src/main/kotlin/PhoneNumber.kt
private const val REQUIRED_LENGTH = 10

class PhoneNumber(input: String) {
    val number: String

    init {
        val digits = input.filter { it.isDigit() }.let {
            if (it.length == 11 && it[0] == '1') it.drop(1) else it
        }

        require(digits.length == REQUIRED_LENGTH) { "Phone number $input is invalid: length is not $REQUIRED_LENGTH" }
        require(digits[0] !in "01") { "Phone number $input is invalid: area code may not start with 0 or 1" }
        require(digits[3] !in "01") { "Phone number $input is invalid: exchange code may not start with 0 or 1" }

        number = digits
    }
}
